### PR TITLE
Add docs for making backends

### PR DESCRIPTION
**Pull request by tjpalmer**
- Leave some things TODO, but get most of the docs on to the site

### CODE REVIEW

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 1]
```diff
@@ -0,0 +1,586 @@
+# So you want to write a Temper backend
+
+## Notice
+
+Backends in the official Temper repository use the same licensing as Temper
+itself. If you want to make a backend with different licensing, you'll need to
+store the code separately from official Temper. We provide information later on
+how to support backends dynamically with the `temper` cli, without maintaining a
+full fork of Temper.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:9]: One of the requested additions.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 2]
```diff
@@ -0,0 +1,586 @@
...
+can copy that which would help you get started.
+
+If your language is a dynamic language, like JavaScript or Python, one of those
+might serve as a template.
+
+If your language is a statically typed, object oriented language that allows at
+most one public class per source file, leaning on the existing Java backend
+might simplify tasks.
+
+### External modules
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:109]: Another requested addition.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 3]
```diff
@@ -0,0 +1,586 @@
...
+  related language) type notations
+
+That comment ends up in the [reference documentation](https://temperlang.github.io/tld/reference/target-languages/#javascript-backend),
+and yours should too once it stabilizes.
+
+## Write an out-grammar
+
+An \*.out-grammar file defines Kotlin classes: one for each kind of output tree
+node, and details on how to "un parse" them back into source code. See
+[out-grammar reference](out-grammar.md) for more details.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:211]: Added this reference.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 4]
```diff
@@ -0,0 +1,586 @@
...
+(TmpL) trees, which are what most backends process to produce their target
+language's output trees.
+
+[Out grammar file format by example](https://github.com/temperlang/temper-prepublic/blob/e39be5d6268a9727b528d89d7094431739632170/kcodegen/src/commonMain/kotlin/lang/temper/kcodegen/outgrammar/README.md#output-grammar-file-format-by-example)
+explains the ins and outs of the file format.
+
+In brief, alternation definitions like the below correspond to a Kotlin
+interface.
+
+```temper inert
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:347]: There's some in common between out-grammar and temper, so I labeled a lot of out-grammar in both files as temper (but inert).

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 5]
```diff
@@ -0,0 +1,586 @@
...
+| ---- | ---- | ---- | ---- | ---- |
+| Render Test | How output trees convert to code | Produce a tree, assert its string form | The out-grammar file changes extensively | Precedence, missing spaces between tokens, and token insertion problems like ASI |
+| Backend Test | The translation of carefully chosen, small Temper programs | Comparing translator output to expected output | Translator changes the trees it produces | Translation corner cases, naming conflicts |
+| Functional Test | Semantics of a suite of carefully chosen Temper programs curated by the Temper core team | Comparing standard output from compiling and running the program using the target language toolchain | The way programs are compiled and launched breaks or the target language has breaking changes | Mis-translations |
+| Connection tests | That hand-written code in the target language can connect to a translated Temper library and get the right results | By writing tests in the target language that depend on translated Temper libraries | The backend changes the promises it makes regarding API translation | Backwards compatibility problems and poor stability of translated APIs |
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:409]: There was a request maybe for a list of connections. I sloppily pulled a list of connected methods from CSharpSupportNetwork. I got 147 of them. I know better would be something automated from the source material. But in any case, we'd want separate references for these things. Too much for here.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 6]
```diff
@@ -0,0 +1,586 @@
...
+Kotlin has a [TODO(\"\$tree\")](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-t-o-d-o.html)
+function that lets you stub out translation paths you're not ready to implement
+yet.
+
+## Lifecycle of a Backend
+
+TODO: Talk about PreAnalysis and metadata, getting your ducks in a row before
+translation, and cross-library dependencies. There are comments for each of the
+lifecycle methods. Maybe turn those into snippets and have a grouping snippet
+that embeds them each.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:500]: I think it's ok to leave these TODOs in place right now. I added some of Mike's additional fresh commentary on possible strategies. But there's enough good material that I'd like to get some out sooner than later.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 7]
```diff
@@ -0,0 +1,586 @@
...
+
+## Support Networks connect Temper builtins to support code
+
+TODO: Flesh out SupportNetwork comments and embed them strategically to produce
+docs on SupportNetwork and what each piece does.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:505]: Also pulled this from Mike's new comment on potential plans.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 8]
```diff
@@ -0,0 +1,586 @@
...
+
+Note that using code from or referencing an existing backend might simplify the
+effort here and in other aspects of writing a backend.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:508]: This was a requested addition. I think I captured the intent?

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 9]
```diff
@@ -0,0 +1,586 @@
...
+`fun translateExpression(expr: TmpL.Expression)` you can write `when (expr) {}`
+and the JetBrains IDE "quick fix" command will happily add all the variants for
+`expr` to the body.
+
+If you leave everything as *TODO()* above, you can pick a functional test and
+see what breaks and fill out translation paths as needed.
+
+Here's an order of functional tests based on an ordre that worked for a recent
+backend effort. However, new functional tests were added since effort, so the
+new tests have been inserted into this list based on estimated fit.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:524]: I adjusted this description slightly and added missing funtests below in places that I think seem maybe ok.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 10]
```diff
@@ -0,0 +1,586 @@
...
+1. AlgosHelloWorld
+1. AlgosFibonacci, TypesIntBasics
+1. AlgosHelloFromClassToTop, AlgosHelloWorldObject
+1. ClassesCallOverrideFromSubtype, ClassesDirectGetter,
+   ClassesInheritedGetter, InterfacesPureVirtual
+1. ControlFlowIfReturn, ControlFlowLoopReenterable, ControlFlowLoops
+1. ControlFlowBubble
+1. CastsAsExpr
+1. FunctionsSimpleLocals, SemanticsMutuallyReferencingTypes
+1. SemanticsTypeCheckedLocals (typically unsupported for static typing)
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:535]: I added a comment on this one.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 11]
```diff
@@ -0,0 +1,586 @@
...
+1. TypesIntLimits, TypesIntShifty
+1. TypesFloatBasics, TypesFloatOps
+1. ImportsFunctions, ImportsValues
+1. NamesNonascii
+1. FunctionsConstructorCallbacks
+1. AlgosMyersDiff, SemanticsBroken, TypesDenseBitVector, TypesDeque
+1. TypesMap
+1. ClassesStaticProperties, ClassesStaticPropertiesScope
+1. ControlFlowActorRun, ControlFlowAsync
+1. TestingAsserts (here down requiring `std`)
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:558]: And on this one.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md [hunk 12]
```diff
@@ -0,0 +1,586 @@
...
+grammar / formatting test
+
+functional test
+
+output grammar
+
+support code
+
+*target language*: a programming language that's a target of translation, such
+as performed by a Temper backend.
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/backend.md:586]: Some of these were sentence fragments. I didn't add missing definitions, but I did try to fill out incomplete ones a little.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md [hunk 13]
```diff
@@ -0,0 +1,651 @@
...
+
+Output grammar files should end with `.out-grammar` so that they are
+automatically found by the code generator task.
+
+All token conventions are the same as those for the Temper language including
+string and C-style comment syntax.
+
+```temper inert
+// This is a comment
+```
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md:57]: Lots of `temper inert` here also for out-grammar code. If we add some pygments support for out-grammar, we could switch.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md [hunk 14]
```diff
@@ -0,0 +1,651 @@
...
+  type is `TopLevels`
+- the `*` is a Kleene star, and means zero-or-more as in regular expression
+  syntax
+- the `"\n"` after the star is the *joiner string*.  When formatting a
+  *Program*, the formatter will insert a newline between each pair of adjacent
+  *TopLevel*s.
+
+After processing this, the generated Kotlin code will have something like
+
+```kt
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md]: I also added `kt` for kotlin code.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md [hunk 15]
```diff
@@ -0,0 +1,651 @@
...
+corresponding Kotlin type will be `MutableList<TopLevel>`.
+
+```temper inert
+ReturnStatement ::= "return" & (expression%Expression || ()) & ";";
+```
+
+Here we have a definition of a *ReturnStatement* node type which uses `||` to
+represent two possible syntaxes
+
+```js
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md:203]: This is technically generated backend code, and I just said `js` for a stand-in.

File: build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+The default expression is direct Kotlin code, so it's in backticks.
+
+### Custom computed properties
+
+Sometimes it's useful to have a computed property in your node type.
+
+In a nested `if` statement, we might not want to put curly brackets around the
+`else` clause if it is also an `if` statement to produce output like
+
+```js
```
Comment (tjpalmer) [inline build-user-docs/skeletal-docs/temper-docs/docs/reference/out-grammar.md:465]: More stand-in for generated code.

File: kcodegen/src/commonMain/kotlin/lang/temper/kcodegen/outgrammar/README.md [hunk 17]
```diff
@@ -3,557 +3,6 @@
 Translator from an output grammar DSL to Kotlin classes meant to
 assist Temper backends in constructing output ASTs.
 
-## How this fits in
+See [output grammar documentation][out-grammar] for help with things here.
```
Comment (tjpalmer) [inline kcodegen/src/commonMain/kotlin/lang/temper/kcodegen/outgrammar/README.md:6]: Left this link as discussed.

**Review by mikesamuel — APPROVED** [review]
